### PR TITLE
Refactor(rust): Improve yaml-parser handling for block scalars

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Improve yaml-parser handling for block scalars

## Related Issue(s)

Fixes a bug where this yaml produces indentation and unmatched quotes errors:

```yaml
key: >-
  'foo
```

Also improves performance significantly.

## Component(s) name

rust

## Proposed changes

- Move block scalar handling to the lexer to it can gather the text without stumbling on other token types.
  - It was an issue to keep this in the emitter since lookaheads would trigger tokenization of the content of the block scalar without understanding the context and different rule-set.

## How to test

- Added many tests for various corner cases
- Reran benchmarks and updated doc

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from main before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.arista.com/devel/docs/contribution/overview.html) document.
- [x] I have updated testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: rust/yaml-parser/tests/edge_cases.rs [hunk 1]
```diff
@@ -805,6 +805,191 @@ fn test_block_scalar_chomping() {
...
+            .as_node()
+            .trailing_comment()
+            .map(|comment| comment.text.as_ref()),
+        Some(" header"),
+    );
+}
+
+#[test]
+fn test_unicode_block_scalar_content_and_header_comment() {
+    let docs = parse_ok("mý🔑: >- # héader ☃\n  café 'quote'\n  emoji 😀\n");
```
Comment (gmuloc) [inline rust/yaml-parser/tests/edge_cases.rs:990]: ❤️

File: rust/yaml-parser/tests/edge_cases.rs [hunk 2]
```diff
@@ -805,6 +805,191 @@ fn test_block_scalar_chomping() {
     }
 }
 
+#[test]
+fn test_folded_block_scalar_quotes_are_content() {
```
Comment (gmuloc) [inline rust/yaml-parser/tests/edge_cases.rs:867]: not sure if we are adding this here but look slike we are not testing empty lines before first content when we are tracking them

File: rust/yaml-parser/src/lexer/block_scalars.rs
Comment (gmuloc) [inline rust/yaml-parser/src/lexer/block_scalars.rs:1]: suggested extra tests for regression testing



  - Tabs around block scalar indentation
```
  key: |
        foo
```
  and:
```
  key: |
        foo
```
  The helper counts only spaces as indent and treats tabs as content or invalid depending on position. That is intentional-looking, but it is a classic YAML edge case.

  - Indented document-marker-looking content
```
  key: |-
    ...
```
  Only zero-indented ... / --- should terminate the scalar. The helper checks indent == 0, so this should be fine, but it’s a useful regression case.

  - Header comment plus following sibling
```
  key: | # header
    value
  next: value
```
  The header comment is queued as a pending token while the block scalar token is returned first. The new comment tests cover attachment, but a sibling-after-scalar case would also check pending token ordering.

  - Explicit indent in nested contexts
```
  outer:
    key: |1
     value
```

all of these are handled correctly by the current PR code.

File: rust/yaml-parser/src/lexer/block_scalars.rs [hunk 3]
```diff
@@ -91,4 +136,335 @@ impl<'input> Lexer<'input> {
...
+                    self.add_error(ErrorKind::InvalidIndentation, line_start_span);
+                }
+                self.pending_tokens.push_back(super::RichToken::new(
+                    Token::LineStart(crate::span::usize_to_indent(indent)),
+                    line_start_span,
+                ));
+                self.byte_pos = line_start;
+                break;
+            }
+            if let Some(ci) = content_indent
```
Comment (gmuloc) [inline rust/yaml-parser/src/lexer/block_scalars.rs:237]: this long functions would benefit from comments to explain the decisions taken (here indent to small, exist the block and move on to next line)

**Review by gmuloc — COMMENTED** [review]
overall ok with the PR - adding a couple of tests suggestion that may be in the YAML suite but would help understand locally what we are doing

File: rust/yaml-parser/src/lexer/block_scalars.rs
Comment (ClausHolbechArista) [reply rust/yaml-parser/src/lexer/block_scalars.rs:1]: added

File: rust/yaml-parser/tests/edge_cases.rs (hunk 2 above)
Comment (ClausHolbechArista) [reply rust/yaml-parser/tests/edge_cases.rs:867]: added tests, from your comment it looks like you are expecting us to trim the leading empty lines?

File: rust/yaml-parser/src/lexer/block_scalars.rs [hunk 4]
```diff
@@ -91,4 +136,366 @@ impl<'input> Lexer<'input> {
...
+                }
+            }
+
+            let ci = content_indent.unwrap_or(indent);
+            let extra_indent = indent.saturating_sub(ci);
+            let has_non_whitespace = line_text.bytes().any(|byte| !matches!(byte, b' ' | b'\t'));
+            // A tab in the visible prefix is content only after the content
+            // indent; before that it is an invalid indentation character.
+            if has_tab_in_prefix && indent < content_indent.unwrap_or(min_auto_indent) {
+                self.add_error(ErrorKind::InvalidIndentation, line_start_span);
```
Comment (gmuloc) [inline rust/yaml-parser/src/lexer/block_scalars.rs]: this is uncoverted according to codecov and this looks to be unreachable.

File: rust/yaml-parser/tests/edge_cases.rs (hunk 2 above)
Comment (gmuloc) [reply rust/yaml-parser/tests/edge_cases.rs:867]:
```
# adding \n for correctness
key: |\n
    \n
  value\n
```

I think this should report invalid indentation but checking codecov ( :) )  I see it is already tested and trqcked it down to `W9L4` from the main suite and it is triggering the InvalidIndentation. So we can resolve this. Thanks for the extra tests

File: rust/yaml-parser/src/lexer/block_scalars.rs (hunk 4 above)
Comment (ClausHolbechArista) [reply rust/yaml-parser/src/lexer/block_scalars.rs]: removed as indeed it is caught earlier.

**Review by gmuloc — APPROVED** [review]
